lib/matplotlib/backend_bases.py

```diff
@@ -3195,3 +3205,508 @@ def zoom(self, *args):
     def set_history_buttons(self):
         """Enable or disable back/forward button"""
         pass
+
+
+class ToolEvent(object):
+    """Event for tool manipulation (add/remove)"""
+    def __init__(self, name, sender, tool, data=None):
+        self.name = name
+        self.sender = sender
+        self.tool = tool
+        self.data = data
+
+
+class ToolTriggerEvent(ToolEvent):
+    """Event to inform  that a tool has been triggered"""
+    def __init__(self, name, sender, tool, canvasevent=None, data=None):
+        ToolEvent.__init__(self, name, sender, tool, data)
+        self.canvasevent = canvasevent
+
+
+class NavigationMessageEvent(object):
+    """Event carrying messages from navigation
+
+    Messages usually get displayed to the user by the toolbar
+    """
+    def __init__(self, name, sender, message):
+        self.name = name
+        self.sender = sender
+        self.message = message
+
+
+class NavigationBase(object):
+    """Helper class that groups all the user interactions for a FigureManager
+
+    Attributes
+    ----------
+    manager: `FigureManager` instance
+    keypresslock: `LockDraw` to know if the `canvas` key_press_event is
+     locked
+    messagelock: `LockDraw` to know if the message is available to write
+    """
+
+    def __init__(self, manager):
+        self.manager = manager
+        self.canvas = manager.canvas
+
+        self._key_press_handler_id = self.canvas.mpl_connect(
+            'key_press_event', self._key_press)
+
+        self._tools = {}
+        self._keys = {}
+        self._toggled = {}
+        self._callbacks = cbook.CallbackRegistry()
+
+        # to process keypress event
+        self.keypresslock = widgets.LockDraw()
+        self.messagelock = widgets.LockDraw()
+
+    def mpl_connect(self, s, func):
```

Comment:
It doesn't shadow at the code level, but it shadows at the user-brain-space level.  I view the consistent interface as a negative as `ToolEvent` and `CanvasEvent` objects are not interchangable.  I suggest calling it `nav_connect` to make clear that it is a different set of call-backs.